rfcs/0123-flake-names.md

```diff
@@ -0,0 +1,89 @@
+---
+feature: flake-names
+start-date: 2022-03-12
+author: Anselm Schüler
+co-authors: None
+shepherd-team: None
+shepherd-leader: None
+related-issues: None
+---
+
+# Summary
+[summary]: #summary
+
+Flakes can declare the field `name`.  
+It represents the name of the flake.  
+The derivations for a flake are no longer called `source`, but use the flake name.
+
+# Motivation
+[motivation]: #motivation
+
+Flake-centric workflows often end up with a lot of derivations named “source”, and it’s difficult to navigate this.
```

Comment:
That’s probably worth a bit more explanations: for most intents and purposes, the store is supposed to be an opaque thing, and not meant to be directly “navigated”.

Now it is a very leaky abstraction, so my remark probably doesn’t really hold in practice, but it would be good to make this explicit (maybe show some concrete examples of where having all the sources be called “source” is an issue)

Reply:
It's an issue if you want to know what a nix command is doing at the moment. It often just displays “building XXXX-source”

Reply:
I might be wrong, but that exact message shouldn’t happen (not because of flakes at least). This message happens when Nix builds a derivation, but flake sources aren’t derivations, they are directly instantiated during the evaluation (with `builtins.fetchTree`) using a different code path with amongst other things a more informative display.

What you might get though is something like “fetching source from https://cache.nixos.org” (because although they aren’t derivations, they can be substituted from the binary cache). Probably we could improve the message in that case.

Reply:
Oh, that’s good to know, sorry for not knowing that.

Reply:
About the store path being meant to be opaque: In my opinion that’s just practically not true at all, and the fact that derivation names exist at all proves it. If anything, this is useful for people that want to spelunk in the depths of what Nix does.

I also regularly use store paths to figure out what package or other source provides some binary in my profile. I have a shell function `goto-bin` for going into the directory of the real path of an executable, and I sometimes skim `realpath` output to see a derivation name.